<a href="https://colab.research.google.com/github/prathameshmahamuni001/Sentiment-analysis-twitter-data/blob/main/Sentiment_analysis_twitter_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
# install kaggle library
! pip install kaggle

In [16]:
#upload kaggle.json
#configuring the path of Kaggle. json file
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/   #
!chmod 600 ~/.kaggle/kaggle.json

In [17]:
#import karo
!kaggle datasets download -d kazanova/sentiment140

sentiment140.zip: Skipping, found more recently modified local copy (use --force to force download)


In [18]:
#extracting the compressed dataset

from zipfile import ZipFile
dataset = '/content/sentiment140.zip'

with ZipFile(dataset, 'r') as zip:
  zip.extractall()
  print('The dataset is extracted')

The dataset is extracted


Import dependencies

In [19]:
import numpy as np
import pandas as pd
import re    #regular expression
from nltk.corpus import stopwords   #natural language tool kit
from nltk.stem.porter import PorterStemmer  #used for trimming the word ig??
from sklearn.feature_extraction.text import TfidfVectorizer  #textual data to numerical data keliye ye kiya
from sklearn.model_selection import train_test_split  #split og data into train and test
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [20]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [21]:
# printing stopwords in English
print(stopwords.words('english'))
#ye sabhi words ko nikaalna hai, dont need it for the project

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

Data processing

In [22]:
#load data from csv
twitter_data = pd.read_csv('/content/training.1600000.processed.noemoticon.csv' , encoding = 'ISO-8859-1')

In [23]:
#checking the number of rows and columns
twitter_data.shape

(1599999, 6)

In [24]:
twitter_data.head()

0  1467810369  Mon Apr 06 22:19:45 PDT 2009  NO_QUERY _TheSpecialOne_  \
0  0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY   scotthamilton   
1  0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY        mattycus   
2  0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY         ElleCTF   
3  0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY          Karoli   
4  0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY        joy_wolf   

  @switchfoot http://twitpic.com/2y1zl - Awww, that's a bummer.  You shoulda got David Carr of Third Day to do it. ;D  
0  is upset that he can't update his Facebook by ...                                                                   
1  @Kenichan I dived many times for the ball. Man...                                                                   
2    my whole body feels itchy and like its on fire                                                                    
3  @nationwideclass no, it's not behaving at all....                                                                   
4                      @Kwesidei not the whole crew

In [25]:
#naming the columns and reading the dataset again, kyuki first 5 rows print karte waqt, 1st row ka data top heading me dikh raha tha which is a glich
#glitch ko thik krne ke liy kiya

column_names = ['target', 'id', 'date', 'flag', 'user', 'text']
#load data from csv
twitter_data = pd.read_csv('/content/training.1600000.processed.noemoticon.csv' , names = column_names, encoding = 'ISO-8859-1')

In [26]:
#checking the number of rows and columns
twitter_data.shape

(1600000, 6)

In [27]:
twitter_data.head()


target          id                          date      flag  \
0       0  1467810369  Mon Apr 06 22:19:45 PDT 2009  NO_QUERY   
1       0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY   
2       0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
3       0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4       0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   

              user                                               text  
0  _TheSpecialOne_  @switchfoot http://twitpic.com/2y1zl - Awww, t...  
1    scotthamilton  is upset that he can't update his Facebook by ...  
2         mattycus  @Kenichan I dived many times for the ball. Man...  
3          ElleCTF    my whole body feels itchy and like its on fire   
4           Karoli  @nationwideclass no, it's not behaving at all....

In [28]:
#counting the number of missing values in the dataset
twitter_data.isnull().sum()

target    0
id        0
date      0
flag      0
user      0
text      0
dtype: int64

In [29]:
#checking distribution of target column (0:negative, 2:neutral, 4:positive)
#number of negative and positive emotion datapoints are equal, this distribution has to be equal,
#if it isn't, then ML model is gonna have problem
twitter_data['target'].value_counts()

target
0    800000
4    800000
Name: count, dtype: int64

Lets convert the target values

In [30]:
#now change karenge labels, it was 0 and 4 for neg and pos, we'll make it 0 and 1 for neg and positive emotion
#first convert target '4' to '1'

In [31]:
twitter_data.replace({'target':{4:1}}, inplace=True)

In [32]:
#checking distribution of target column (0:negative, 2:neutral, 4:positive)

twitter_data['target'].value_counts()
#now 0 means negative and 1 positive emotion text

target
0    800000
1    800000
Name: count, dtype: int64

0 ------> Negative



1 ------> Positive

*Stemming*

Stemming is the process of reduction of word to its root word, word ko chota banana

Example : actor, actress, acting = act

In [33]:
port_stem = PorterStemmer()

In [34]:
def stemming(content):
  stemmed_content = re.sub('[^a-zA-Z]',' ', content)  #from dataset, this will remove all the words which contains any symbol like @,# and will only keep the words a-z and A-Z for stemming.
  stemmed_content = stemmed_content.lower()  #sab letters ko lower kardo
  stemmed_content = stemmed_content.split()  #list bann jayega after splitting
  stemmed_content = [port_stem.stem(word) for word in stemmed_content if not word in stopwords.words('english')]
  stemmed_content = ' '.join(stemmed_content)

  return stemmed_content

In [35]:
twitter_data['stemmed_content'] = twitter_data['text'].apply(stemming)  #50 minutes to complete this execution

In [36]:
twitter_data.head()   #As you can see in stemmed_content, text me jitne bhi symbols the faltu ke hatt gaye, letters small hogye, split hogaye
#all because of def stemming in line 27th

target          id                          date      flag  \
0       0  1467810369  Mon Apr 06 22:19:45 PDT 2009  NO_QUERY   
1       0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY   
2       0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
3       0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4       0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   

              user                                               text  \
0  _TheSpecialOne_  @switchfoot http://twitpic.com/2y1zl - Awww, t...   
1    scotthamilton  is upset that he can't update his Facebook by ...   
2         mattycus  @Kenichan I dived many times for the ball. Man...   
3          ElleCTF    my whole body feels itchy and like its on fire    
4           Karoli  @nationwideclass no, it's not behaving at all....   

                                     stemmed_content  
0  switchfoot http twitpic com zl awww bummer sho...  
1  upset updat facebook text might cri result sch...  
2  kenichan dive mani time ball manag save rest g...  
3                    whole bodi feel itchi like fire  
4                      nationwideclass behav mad see

In [37]:
print(twitter_data['stemmed_content'])

0          switchfoot http twitpic com zl awww bummer sho...
1          upset updat facebook text might cri result sch...
2          kenichan dive mani time ball manag save rest g...
3                            whole bodi feel itchi like fire
4                              nationwideclass behav mad see
                                 ...                        
1599995                           woke school best feel ever
1599996    thewdb com cool hear old walt interview http b...
1599997                         readi mojo makeov ask detail
1599998    happi th birthday boo alll time tupac amaru sh...
1599999    happi charitytuesday thenspcc sparkschar speak...
Name: stemmed_content, Length: 1600000, dtype: object


In [38]:
print(twitter_data['target'])

0          0
1          0
2          0
3          0
4          0
          ..
1599995    1
1599996    1
1599997    1
1599998    1
1599999    1
Name: target, Length: 1600000, dtype: int64


In [39]:
#seperate data n label
X = twitter_data['stemmed_content'].values
Y = twitter_data['target'].values

In [40]:
print(X)

['switchfoot http twitpic com zl awww bummer shoulda got david carr third day'
 'upset updat facebook text might cri result school today also blah'
 'kenichan dive mani time ball manag save rest go bound' ...
 'readi mojo makeov ask detail'
 'happi th birthday boo alll time tupac amaru shakur'
 'happi charitytuesday thenspcc sparkschar speakinguph h']


In [41]:
print(Y)

[0 0 0 ... 1 1 1]


Splitting the data to training data and test data

In [42]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [43]:
print(X.shape, X_train.shape, X_test.shape)

(1600000,) (1280000,) (320000,)


In [44]:
print(X_train)

['watch saw iv drink lil wine' 'hatermagazin'
 'even though favourit drink think vodka coke wipe mind time think im gonna find new drink'
 ... 'eager monday afternoon'
 'hope everyon mother great day wait hear guy store tomorrow'
 'love wake folger bad voic deeper']


In [45]:
print(X_test)

['mmangen fine much time chat twitter hubbi back summer amp tend domin free time'
 'ah may show w ruth kim amp geoffrey sanhueza'
 'ishatara mayb bay area thang dammit' ...
 'destini nevertheless hooray member wonder safe trip' 'feel well'
 'supersandro thank']




```
# This is formatted as code
```

CONVERTING THE TEXTUAL DATA TO NUMERICAL DATA --> FEATURES EXTRACTION

In [46]:
vectorizer = TfidfVectorizer()
X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

In [47]:
print(X_train)

  (0, 443066)	0.4484755317023172
  (0, 235045)	0.41996827700291095
  (0, 109306)	0.3753708587402299
  (0, 185193)	0.5277679060576009
  (0, 354543)	0.3588091611460021
  (0, 436713)	0.27259876264838384
  (1, 160636)	1.0
  (2, 288470)	0.16786949597862733
  (2, 132311)	0.2028971570399794
  (2, 150715)	0.18803850583207948
  (2, 178061)	0.1619010109445149
  (2, 409143)	0.15169282335109835
  (2, 266729)	0.24123230668976975
  (2, 443430)	0.3348599670252845
  (2, 77929)	0.31284080750346344
  (2, 433560)	0.3296595898028565
  (2, 406399)	0.32105459490875526
  (2, 129411)	0.29074192727957143
  (2, 407301)	0.18709338684973031
  (2, 124484)	0.1892155960801415
  (2, 109306)	0.4591176413728317
  (3, 172421)	0.37464146922154384
  (3, 411528)	0.27089772444087873
  (3, 388626)	0.3940776331458846
  (3, 56476)	0.5200465453608686
  :	:
  (1279996, 390130)	0.22064742191076112
  (1279996, 434014)	0.2718945052332447
  (1279996, 318303)	0.21254698865277746
  (1279996, 237899)	0.2236567560099234
  (1279996, 2910

In [48]:
print(X_test)

  (0, 420984)	0.17915624523539803
  (0, 409143)	0.31430470598079707
  (0, 398906)	0.3491043873264267
  (0, 388348)	0.21985076072061738
  (0, 279082)	0.1782518010910344
  (0, 271016)	0.4535662391658828
  (0, 171378)	0.2805816206356073
  (0, 138164)	0.23688292264071403
  (0, 132364)	0.25525488955578596
  (0, 106069)	0.3655545001090455
  (0, 67828)	0.26800375270827315
  (0, 31168)	0.16247724180521766
  (0, 15110)	0.1719352837797837
  (1, 366203)	0.24595562404108307
  (1, 348135)	0.4739279595416274
  (1, 256777)	0.28751585696559306
  (1, 217562)	0.40288153995289894
  (1, 145393)	0.575262969264869
  (1, 15110)	0.211037449588008
  (1, 6463)	0.30733520460524466
  (2, 400621)	0.4317732461913093
  (2, 256834)	0.2564939661498776
  (2, 183312)	0.5892069252021465
  (2, 89448)	0.36340369428387626
  (2, 34401)	0.37916255084357414
  :	:
  (319994, 123278)	0.4530341382559843
  (319995, 444934)	0.3211092817599261
  (319995, 420984)	0.22631428606830145
  (319995, 416257)	0.23816465111736276
  (319995, 3

Training the ML model


LOGISTIC REGRESSION  //Why used this? to classify the data into classes or subclasses

In [49]:
model = LogisticRegression(max_iter=1000)  #1000 number of times the model can go through the data

In [50]:
model.fit(X_train,Y_train)

LogisticRegression(max_iter=1000)

Model evaluation

Accuracy score

In [51]:
#accuracy score on the training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(Y_train,X_train_prediction)

In [52]:
print('Accuracy score on training data :', training_data_accuracy)

Accuracy score on training data : 0.81018984375


In [53]:
#accuracy score on the training data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(Y_test,X_test_prediction)

In [54]:
print('Accuracy score on testing data :', test_data_accuracy)

Accuracy score on testing data : 0.7780375


Model accuracy = 77.8%

TRAINING OVER !!

Saving the trained model

In [55]:
import pickle

In [56]:
filename = 'trained_model.sav'
pickle.dump(model,open(filename, 'wb'))

Using saved model for future prediction

In [57]:
#loading the saved model
loaded_model = pickle.load(open('/content/trained_model.sav', 'rb'))

In [58]:
X_new = X_test[200]
print(Y_test[200])

prediction = model.predict(X_new)
print(prediction)

if (prediction[0] == 0):
  print('Negative Tweet')

else:
  print('Positive Tweet')

1
[1]
Positive Tweet


In [59]:
X_new = X_test[3]
print(Y_test[3])

prediction = loaded_model.predict(X_new)
print(prediction)

if (prediction[0] == 0):
  print('Negative Tweet')

else:
  print('Positive Tweet')

0
[0]
Negative Tweet


In [59]:
from flask import Flask, render_template, request
import pickle
import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer

app = Flask(__name__)

# Load the trained model
model = pickle.load(open('trained_model.sav', 'rb'))

# Load the TF-IDF vectorizer used during training
vectorizer = pickle.load(open('tfidf_vectorizer.sav', 'rb'))

# Preprocessing functions
def preprocess(text):
    # Remove special characters and digits
    text = re.sub(r'\W', ' ', text)
    text = re.sub(r'\d', ' ', text)
    # Convert to lowercase
    text = text.lower()
    # Tokenization
    text = text.split()
    # Remove stopwords and perform stemming
    ps = PorterStemmer()
    text = [ps.stem(word) for word in text if not word in set(stopwords.words('english'))]
    # Join tokens back into string
    text = ' '.join(text)
    return text

# Route for home page
@app.route('/')
def home():
    return render_template('index.html')

# Route for sentiment analysis
@app.route('/analyze', methods=['POST'])
def analyze():
    if request.method == 'POST':
        # Get text from form submission
        text = request.form['text']
        # Preprocess input text
        preprocessed_text = preprocess(text)
        # Vectorize preprocessed text
        vectorized_text = vectorizer.transform([preprocessed_text])
        # Predict sentiment
        sentiment = model.predict(vectorized_text)[0]
        # Map sentiment label to readable format
        sentiment_label = "Positive" if sentiment == 1 else "Negative"
        # Render result template with sentiment prediction
        return render_template('result.html', sentiment=sentiment_label)

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
from google.colab import drive
drive.mount('/content/drive')